### ISSUE TITLE
Add keyboard handlers

### ISSUE THREAD

**Issue by doronrosenberg** [maqetta/maqetta#2640]
@JonFerraiolo @jhpedemonte 

Runtime.js would handle the basics for us.  It would provide `registerKeyBinding(keyBinding, pluginAction)`, which we would call to register key bindings.  KeyBinding would be an object that has the accelerator, key and any modifiers (shift for example).

When we parse plugins, like davinci.actionSets, we check if there is keyBinding object.  If it does, we then register that keyBinding and its associated pluginAction object as a global key binding.

However, since we use iframes for certain editors, every editor that does so would add a general key listener to the iframe document and then pass that event object to `Runtime.js::handleKeyEvent(Event e)` which would check if that key binding is registered and if so call it.

For editor-specific event handling, that would be handled by the appropriate class and not use plugins to simplify things.   Events that are not handled by this editor specific code would then get passed on to Runtime.js::handleKeyEvent``` to be handled globally.

The one problem I see is do we want for example users to save using ctrl-s even if focus is outside the editor (say we are in the prop panel)?  If so, then global key events would have to be passed to the current editor to be handled as well, and we would have to decide who gets first stab at global events - the editor or Runtime's code?

**Comment by JonFerraiolo** [comment]
> The one problem I see is do we want for example users to save using ctrl-s even if focus is outside the editor (say we are in the prop panel)?

This is one of those things that we might have to implement one way, try out, and then adapt.

But in general, I would think that Ctrl-S should do a save no matter where the focus is - in an editor, in the properties palette, in the widget filter field, doesn't matter. But in other cases - eg, Ctrl-X/Ctrl-C/Ctrl-V - we should do different things depending on whether focus is in a palette field versus in an editor.

**Comment by jhpedemonte** [comment]
Continuing from what Jon said:  the plan that you have outlined will work well for keyboard bindings which are _global_.

However, how will you handle bindings which only affect one part of Maqetta?  Or (as with cut/copy/paste) there is a binding which can be registered multiple times from different parts?

Probably best to go forward with the known _global_ bindings and get those working well.  Then iterate on top of that code to handle the more complex situations.

**Comment by doronrosenberg** [comment]
@jhpedemonte Non-global bindings would be handled by the widgets themselves - say the outline tree, it would do its own thing and cancel any events it handles.

**Comment by jhpedemonte** [comment]
And if, say, Outline and Files view both register the same keyboard binding, which one gets called?  I'm not sure if "whichever has focus" is an adequate answer, since that always isn't evident to the user.

For example, bring up Sample1.html and make the Outline view visible.  Select the **New** button in the VE -- notice that it is selected in the VE and also in the Outline.  Now click on **New** button in the Outline -- it gets a _focus rectangled_ in addition to the _selection_ rectangle, but the selection in the VE stays the same.  Now select **app.css** in the Files view -- it gets the _focus_ rectangle now, instead of the item in the Outline, but the Outline still shows the _selection_.

Even to me, it would be hard to track which one actually has the focus.  In particular, there are no visual cues that there is focus in the VE, as there are (somewhat) with the trees in the Outline and Files views.

**Comment by doronrosenberg** [comment]
For Outline and Files, they would handle the events themselves, so if Files has the "real" focus then it would handle the key press.

Probably a question for the UX team? @JonFerraiolo

**Comment by JonFerraiolo** [comment]
@jmockler @peller

One thing to consider upfront is that we need to keep accessibility in mind. The reason that icons take the focus is mostly for accessibility reasons, particularly to allow the user interface to be navigated via the keyboard (or voice actions).

I suspect that there isn't going to be a simple or singular rule for how to deal with focus issues and accelerator keys. There will be some accelerator keys that we want Maqetta to handle, some we will want dojo to handle, others that we will want the browser to handle, and the decision about which accelerators go to Maqetta vs dojo vs browser are likely to be context-dependent in various cases.

A proposed overly-simplified answer (that undoubtedly will have to have various exceptions in practice) is that we should allow dojo and browser to deal with focus control and accelerator keys in general, but that we override their default keyboard actions in selected cases when essential for reasonable user experience. 

For example, I'm pretty sure we want Ctrl-S to do a document save on the current editor no matter whether the browser focus is in the visual page editor, source editor or in one of the palettes. I expect it's just common sense. We just need to ask ourselves, from a user perspective, what would a user expect to happen when they issue an accelerator key in different situations. For example, Ctrl-C immediately after selecting in the visual page editor should copy the current selection, but Ctrl-C after selecting some text in field in the Properties palette should copy that selected text.

Note that some keys will be more complicated, such as arrow keys. I have inserted some logic into the code base to force the focus into the visual page editor at various points so that the visual page editor's iframe will receive keyboard events so that arrow keys will move the current selection. I attempted to force focus only in response to actions such as selection that are page editing centric. But note that if you click on an icon in Files palette, focus is now in Files palette, and arrow keys will jump to next/previous icon there.

**Comment by peller** [comment]
Note also the platform issues with at least some accelerators: e.g. Mac command key vs. Windows control key.   Clipboard operations are command-C/X/V on the Mac, and the native ops will tend to work regardless of browser restrictions (though perhaps not in an extensible way, or quite as you'd like them to work!)  Even ctrl-S may be conflict with a browser save-as menu?

Regarding accessibility, I believe there are "standards" around accelerator keys to use for navigation, and last I checked the keyboard spec was still an 'emerging' standard which did not account for all widgets, but Dijit tried to follow it.  With containers, it's not always obvious how to combine navigation within custom widgets vs. across the page hierarchy and not have our own accelerators conflict with that navigation.  Tab usually moves from one widget to another, but arrows are often used within compound widgets like tab containers.  We could consult one of the accessibility folks to see where this stands, but I agree with Jon, we should defer up the application stack where possible, otherwise we'll just have to do our best.

**Comment by doronrosenberg** [comment]
@jhpedemonte So Chrome won't allow use to intercept accel-n, accel-w and accel-t, need to update the google doc.  Perhaps use alt on windows and ctrl on mac?

**Comment by doronrosenberg** [comment]
Full list is Ctrl-PageUp
Ctrl-PageDown
Ctrl-W
Ctrl-F4
Ctrl-T
Ctrl-N
Ctrl-Shift-N

**Comment by JonFerraiolo** [comment]
For W and N, I can't think of anything better than alt on windows and ctrl on mac.

**Comment by jhpedemonte** [comment]
No, I'd rather keep the default accelerator for each platform, rather than have some using default but others using something else.

How about something like this:

New HTML _E_ditor:  {Accel} E [1]
New Generic _E_ditor: {Accel} Shift E
Close: {Accel} Shift W [2]

[1] I noticed that Google Docs doesn't have keyboard bindings for _New File_ or _Close_, which correspond to {Accel}-N and {Accel}-W -- you can only do those actions through menus.  Otherwise, all keyboard bindings use the default OS keyboard accelerator.

[2] Also, we may want to register a listener for closing/navigating away from the page and popup an "are you sure you want to close Maqetta" dialog.  Would prevent people from accidentally closing Maqetta if they press {Accel}-W.

**Comment by doronrosenberg** [comment]
@jhpedemonte accel shift w was close all editors, what do we switch to then? ctrl shift a?

**Comment by doronrosenberg** [comment]
accel shift w also doesn't work well, chrome for example will close itself anyways.

**Comment by doronrosenberg** [comment]
I've checked in my current progress into master.  We've got most of the bindings we want, I'll be adding the test in the next day or so

**Comment by jhpedemonte** [comment]
(Not sure if this should be a separate bug.  Feel free to spin off.)

Found some keyboard binding issues (while testing on Chrome/Mac):
- **Cmd+N** doesn't work.  As mentioned above, we cannot use this binding, but it is listed in test plan.
- Undo/redo should always work on the current VE, even if it doesn't have focus. (Instructions say to click in VE in order for commands to work.)  Only exceptions would be if, for example, focus is in a text area.  (See the VE/source/split view keyboard bindings, which work even if VE doesn't have focus.)
- **delete** key on MacBook does not delete a widget.  Need to do **fn+delete**.

**Comment by doronrosenberg** [comment]
@jhpedemonte what should we do for accel w and chrome - we can't use any combination that includes accel and w.

**Comment by jhpedemonte** [comment]
> @jhpedemonte what should we do for accel w and chrome - we can't use any combination that includes accel and w.  

Some options:
1. **Choose a different letter.**  Given the string "Close Window", either C or W make sense.  We can't use the latter, and the former is used for Copy.  We could take the L from "close", but Accl-L just seems weird (well, any other letter seems weird).
2. **No keyboard binding for close window.**  This is the approach taken by Google Docs, for example.
3. **Use the Meta key instead.**  Since we cannot use Cmd-W (OS X), we could instead use Ctrl-W.  I don't particularly like this option, but it's better than (1).

**Comment by JonFerraiolo** [comment]
My two cents: option 2 - just don't offer an accelerator for this operation.

**Comment by jhpedemonte** [comment]
We can go with (2) for now, and see how it feels...

**Comment by doronrosenberg** [comment]
Ended up going with 2.  Closing this out as we now support everything we want to support, any bugs should be filed separately.
